### PR TITLE
Vulkan generator 2

### PR DESCRIPTION
**Pull request by yalcinmelihyasin**

### CODE REVIEW

File: vulkan_generator/tests/test_types.py [hunk 1]
```diff
@@ -22,8 +22,8 @@
 
 import xml.etree.ElementTree as ET
 
-from vulkan_parser import handle_parser
-from vulkan_parser.types import VulkanHandle, VulkanHandleAlias
+from vulkan_parser import handle_parser, struct_parser
+from vulkan_parser.types import VulkanHandle, VulkanHandleAlias, VulkanStruct, VulkanStructAlias
```
Comment (purvisa-at-google-com) [inline vulkan_generator/tests/test_types.py:26]: Should we be importing all of vulkan_parser.types here? Or if we have to import multiple things in isolation should we import them one per line? ie 

from vulkan_parser.types import VulkanHandle
from vulkan_parser.types import VulkanHandleAlias
...

What we have above is not scalable if we end up importing 12 different classes

File: vulkan_generator/vulkan_utils/parsing_utils.py [hunk 2]
```diff
@@ -14,6 +14,7 @@
 
 """This module contains the utility functions that needed elsewhere while parsing Vulkan XML"""
 
+import os
 import xml.etree.ElementTree as ET
```
[the remark this answers is not in the document]
Comment (yalcinmelihyasin) [reply vulkan_generator/vulkan_utils/parsing_utils.py:18]: The issue here is that `ElementTree` is also a type name.

File: vulkan_generator/tests/test_types.py (hunk 1 above)
Comment (yalcinmelihyasin) [reply vulkan_generator/tests/test_types.py:26]: :+1:

File: vulkan_generator/vulkan_parser/struct_parser.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        #
+        # <member len="enabledLayerCount,null-terminated">const <type>char</type>
+        # * const*      <name>ppEnabledLayerNames</name>
+        #
+        type_attributes = parsing_utils.try_get_text_from_tag(member_element, "member")
+        # some times it's just empty space or endline character
+        if type_attributes is not None:
+            type_attributes = parsing_utils.clean_type_string(type_attributes)
+
+        if type_attributes and len(type_attributes) > 0:
```
[the remark this answers is not in the document]
Comment (purvisa-at-google-com) [reply vulkan_generator/vulkan_parser/struct_parser.py:77]: This is also written in the google open source python style guide:

https://github.com/google/styleguide/blob/gh-pages/pyguide.md

File: vulkan_generator/vulkan_utils/parsing_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (purvisa-at-google-com) [reply vulkan_generator/vulkan_utils/parsing_utils.py:18]: In this case I think it's ok then. But as discussed in person we should avoid doing this in other locations

**Review by purvisa-at-google-com — APPROVED** [review]

File: vulkan_generator/vulkan_parser/struct_parser.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    </member>
+    """
+
+    #  This is a bit of an irregular code because the XML itself has quite irregularities that
+    # makes is hard to parse type and variable easily.
+    #
+    # This is not the code we wanted but it's the code that we needed and it's contained in a
+    # small place so that XML irregularities does not leak into the rest of the code.
+
+    members = {}
```
[the remark this answers is not in the document]
Comment (purvisa-at-google-com) [reply vulkan_generator/vulkan_parser/struct_parser.py:46]: +1

File: vulkan_generator/vulkan_parser/struct_parser.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+        pointers = parsing_utils.try_get_tail_from_tag(member_element, "type")
+        if pointers is not None:
+            pointers = parsing_utils.clean_type_string(pointers)
+            # Add space between "*" and "const"
+            pointers = pointers.replace("const", " const")
+            typename = f"{typename}{pointers}"
+
+        if typename is None:
+            raise SyntaxError(f"No typename found in : {ET.tostring(member_element, 'utf-8')}")
```
[the remark this answers is not in the document]
Comment (yalcinmelihyasin) [reply vulkan_generator/vulkan_parser/struct_parser.py:88]: That's why I wanted to raise exception. I don't know if it's pythonic(please let me know the correct way if it's not) but that's what I was trying to tell here.  If the typename is `None` here it's wrong. Maybe I should check both `None` and empty string. Would it be:

`if not typename:` ? Or something similar.

Basically general principle is, fail as early as possible if something is wrong. Because we know the XML that we are parsing, it shouldn't go wrong and if we update the XML and things changes, then we would know what has changed.

File: vulkan_generator/vulkan_parser/types.py [hunk 6]
```diff
@@ -33,3 +38,42 @@ class VulkanHandle(VulkanType):
...
     """The meta data defines a Vulkan Handle alias"""
     aliased_typename: str
+
+@dataclass
+class VulkanStructMember(VulkanType):
+    """The meta data defines a Vulkan Handle"""
+    variable_name: str
+
+    # Some member variables are static arrays with a default size
+    varible_size: str
```
[the remark this answers is not in the document]
Comment (yalcinmelihyasin) [reply vulkan_generator/vulkan_parser/types.py:48]: 👍

File: vulkan_generator/vulkan_generator.py [hunk 7]
```diff
@@ -15,19 +15,26 @@
 """This is the top level point for Vulkan Code Generator"""
 
 from pathlib import Path
-from pprint import pprint
+from pprint import PrettyPrinter
```
[the remark this answers is not in the document]
Comment (purvisa-at-google-com) [reply vulkan_generator/vulkan_generator.py:18]: No, I think the opposite is true actually, you're not supposed to import single classes from packages?

File: vulkan_generator/tests/test_types.py [hunk 8]
```diff
@@ -23,7 +23,11 @@
 import xml.etree.ElementTree as ET
 
 from vulkan_parser import handle_parser
-from vulkan_parser.types import VulkanHandle, VulkanHandleAlias
+from vulkan_parser import struct_parser
+from vulkan_parser.types import VulkanHandle
+from vulkan_parser.types import VulkanHandleAlias
+from vulkan_parser.types import VulkanStruct
+from vulkan_parser.types import VulkanStructAlias
```
[the remark this answers is not in the document]
Comment (purvisa-at-google-com) [reply vulkan_generator/tests/test_types.py:30]: https://github.com/google/styleguide/blob/gh-pages/pyguide.md#22-imports

The style guide says not to import individual classes and functions. If we are importing multiple things from the same package though, we should use one line per thing I believe. so instead of:

from package import Thing, OtherThing, ThirdThing

we should use:

from package import Thing
from package import OtherThing
from package import ThirdThing

But consider just 

import package

Unless there's a good reason not to

File: vulkan_generator/vulkan_parser/struct_parser.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (paulthomson) [reply vulkan_generator/vulkan_parser/struct_parser.py:88]: Oh this is fine. I guess my point was more: `typename` cannot be `None` according to the type hints. So, if you ever violated this in the future, `mypy` should fail. So, you could rely on the type hints to catch this, instead of using an exception or assert. It would be similar to if the type of `typename` was `std::string` in C++; it cannot be nullptr, so you would not check for this (although you might check that isn't zero length).

Even if you don't want to rely on the type hints, some people would argue that you should use:

```python
assert typename, "Expected typename to have a non-zero value!"
```

here instead of an exception, as `typename` should always be set, regardless of input values (the XML, I guess). I.e. it is guaranteed by the (lack of) control flow above, and by the fact that `parsing_utils.get_text_from_tag` cannot return None, and is not affected by the XML input. Actually, perhaps this is not true because the XML could contain a zero-length string. Regardless, I would be fine with it. On the other hand, some people say that you should not use `assert` outside tests. The Google style seems to allow/encourage asserts for checking "internal correctness", which is what I mean here.
 
If you still want to use an `if`, then yes, I suggest:

```
if not typename:
  raise ...
```

File: vulkan_generator/vulkan_utils/parsing_utils.py [hunk 9]
```diff
@@ -14,14 +14,45 @@
...
-    for child in root:
+    value = try_get_text_from_tag(elem, tag)
+
+    if value is None:
+        # This should not happen
+        raise SyntaxError(f"No {tag} tag found in {ET.tostring(elem, 'utf-8')}")
+
+    return value
+
+def try_get_text_from_tag(elem: ET.Element, tag: str) -> str:
```
[the remark this answers is not in the document]
Comment (paulthomson) [reply vulkan_generator/vulkan_utils/parsing_utils.py:31]: > So if we are returning either a string or None from a function, then the function returns Optional[str]

Yes

> I thought that Optional would be a type, similar to java.

Not really. It is not a class that you can instantiate. It is just used for type hints. `mypy` is pretty clever with Optional:

```python
# x has type Optional[str]

if not x:
  raise ValueError()

# Here, mypy knows x has type `str` (i.e. x definitely is not None)

# Assume `def use_stuff(x: str) -> None`.
# If we did not do the `if` check above,
# then `mypy` would disallow the call below because
# we are passing `Optional[str]` but the function expects `str`.
use_stuff(x)

```

[the remark this answers is not in the document]
Comment (yalcinmelihyasin) [reply vulkan_generator/vulkan_utils/parsing_utils.py]: 👍

File: vulkan_generator/vulkan_parser/struct_parser.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yalcinmelihyasin) [reply vulkan_generator/vulkan_parser/struct_parser.py]: 👍

File: vulkan_generator/vulkan_parser/struct_parser.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yalcinmelihyasin) [reply vulkan_generator/vulkan_parser/struct_parser.py]: 👍

File: vulkan_generator/tests/test_types.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (yalcinmelihyasin) [reply vulkan_generator/tests/test_types.py]: 👍

[the remark this answers is not in the document]
Comment (yalcinmelihyasin) [reply vulkan_generator/tests/test_types.py]:
```
https://github.com/google/styleguide/blob/gh-pages/pyguide.md#22-imports

The style guide says not to import individual classes and functions. If we are importing multiple things from the same package though, we should use one line per thing I believe. so instead of:

from package import Thing, OtherThing, ThirdThing

we should use:

from package import Thing
from package import OtherThing
from package import ThirdThing

But consider just

import package

Unless there's a good reason not to
```

I think this is not wanted.

File: vulkan_generator/vulkan_generator.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (yalcinmelihyasin) [reply vulkan_generator/vulkan_generator.py]: 👍

File: vulkan_generator/vulkan_parser/struct_parser.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yalcinmelihyasin) [reply vulkan_generator/vulkan_parser/struct_parser.py]: 👍

**Review by purvisa-at-google-com — APPROVED** [review]

**Comment by yalcinmelihyasin** [comment]
Linux test failure is irrelevant.
